### DISCUSSION TITLE
Allow Structs Derivations - for adding Non-Field Members

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/csharplang#5890] [Language Ideas]
In our quest for create efficient Extendible Enums, we are doing this now with StructEnums (to avoid making each Enum instance into a class reference).

We want to be able to Derive from a base Struct for purposes of adding more Static members only (not adding any state values to each instance).   

What we want to be able to do is this:

```c#
public struct StructEnum
{
      protected int _index;  // NOTE: "protected" here, not currently allow in Structs.

      // all the infrastructure to make this work nice -- mapping Names to an Index, which then can point to more information.  A flyweight

}


namespace AnotherAssembly
{
   public struct MapLayerName : StructEnum
   {
        static public MapLayerName Roads = new MapLayerName("Roads");
        //.. more static members like this

        // We can add more data to this struct via more static Arrays of data -- accessed by the "_index".
   }
}
```
Now we have easy Java-Like-enums that are extendible, and are also Structs, so very cheap and adding no burden to the GC.

None of this would change the underlying data layout of the "StructEnum".  It would REMAIN a simple "int _index" wrapper.  That's it.

**The Derived struct types could add functionality, but not Fields, making this a feasible request.**

**Comment by HaloFour** [comment]
The runtime forbid extending a struct.

There's no reason to make this a struct.  Given an enum is a small set of instances which would only be allocated once that puts no pressure on the heap or GC.  Depending on the number of members you're much more likely to incur a performance penalty by declaring them as structs.

**Comment by CyrusNajmabadi** [comment]
Struct inheritance discussion is covered here: https://github.com/dotnet/csharplang/issues/524
